### ISSUE TITLE
Question: how to programmatically bootstrap a cookiecutter

### ISSUE THREAD

**Issue by jamesbraza** [cookiecutter/cookiecutter#1476]
* Cookiecutter version: `cookiecutter==1.7.2` (current stable)
* Python version: `Python==3.8.6`

### Description:

I am spooling up a `cookiecutter` repository called `my-cookiecutter` (not the real name), and have it checked out locally.  

`cookiecutter.json`

```json
{
    "foo": "",
}
```

I have another repo called `instance-01` (not the real name) where I want to have pre-supplied parameters for `cookiecutter.json` in a file.

`instance-01-cookiecutter-config.json`

```json
{
    "foo": "bar",
}
```

I would like to run `cookiecutter ../path/to/my-cookiecutter` and somehow take in the `instance-01-cookiecutter-config.json` (instead of having to manually fill out prompts).

I have read through most of the `cookiecutter` docs and many issues, and can't figure out if this is something `cookiecutter` can do.

In other words, how can I programmatically bootstrap a `cookiecutter`?  Is it possible?

**Comment by tpoindessous** [comment]
Hi ! 

I looked for the same thing and it doesn't seem to exist ?

The option `--directory` could be useful but it's looking for a directory inside the remote repo.

Like you, I would use cookiecutter with a remote repo and a local configuration file (cookiecutter.json).

Thanks !

**Comment by tpoindessous** [comment]
I was looking at last commits and I saw this one : 

https://github.com/cookiecutter/cookiecutter/commit/d74b085260248f5ab180f9e582ea973108fa3e2b

Cookiecutter authors added a new option for selecting replay file.

This is not exactly what we want :) but it can be an alternative or the beginning of a new option :)

**Comment by dHannasch** [comment]
Does `--config-file` not do what you want? Is it important that the file be JSON rather than YAML?

**Comment by jamesbraza** [comment]
Thanks for the response @dHannasch !  No the file doesn't have to be JSON, it can be YAML.

I tried this out to see.  It's _halfway_ what I am asking for.
- Pro: through `default_context`, one can override defaults set in `cookiecutter.json`
    - Using terminology from the original post, this means `instance-01-cookiecutter-config.json` --> `instance-01-cookiecutter-config.yaml`
- Pro: through `replay_dir`, one can programmatically output the values fed in (I didn't know about this)

However, this still isn't _programmatic_.  One has to manually input values, even if it's just hitting enter to accept a default.  Do you know of a way to programmatically invoke a `cookiecutter` such that one doesn't have to manually hit enter a bunch of times (accepting defaults)?

**Comment by edesz** [comment]
Hi @jamesbraza, unless I'm misunderstanding the issue, this is exactly what `cookiecutter`'s `--config-file` ([docs](https://cookiecutter.readthedocs.io/en/latest/advanced/user_config.html#user-config-0-7-0)) command line argument is designed to address. This is a YAML file that accepts a `default-context` block consisting of the template parameter's name-value pairs (specified in `cookiecutter.json`) that will be injected as default values when you generate a templated project with `cookiecutter`. So, for example if you had `{"foo": ""}`  in your `cookiecutter.json`, then you could specify `{"foo": "bar"}` as one of the name-value pairs in the `default-context` block of a configuration file.

With this, the template's default values are set to be used. If you now run `cookiecutter --config-file <your-config-file>.yaml`, `cookiecutter` will **still** prompt for input, but now the values from `<your-config-file>.yaml` will be injected as the default values at the corresponding prompts. With this setup, you could use `cookiecutter` in two ways
1. If you generate your template with
    ```
    $ cookiecutter
    ````
    then you would get
    ```
    foo []:
    ```
    as the prompt but you have to manually enter a value before you can press **`Enter`**.
2. On the other hand, if you generate the template with
    ```
    $ cookiecutter --config-file <your-config-file>.yaml
    ````
    then you would get
    ```
    for [bar]:
    ```
    as the prompt. Now, you can simply press `Enter` without needing to supply any value because the value for `foo` will be automatically set to `bar` (which comes from `<your-config-file>.yaml`).

So, what's left to do now, is to inform `cookiecutter` to actually skip all prompts. To avoid any user input, the CLI argument `--no-input` ([docs](https://cookiecutter.readthedocs.io/en/latest/advanced/cli_options.html#cmdoption-cookiecutter-no-input)) is required. This skips any prompts and uses the values from `cookiecutter.json`. However, since a custom configuration file is also being specified, those config file settings are treated like the defaults in `cookiecutter.json` and so those parameters are used in the template.

To summarize all this, here's a simple example of the overall workflow

1. Set up `cookiecutter.json
    ```
    {
        "foo": "",
    }
    ```
2. set up the config file you to automatically specify a value for `foo` when generating a template
    ```
    {
        "foo": "bar",
    }
    ```
3. Use the template to render a project with no user inputs from the CLI
    ```
    $ cookiecutter my-cookiecutter --no-input --config-file /home/<username>/my-cookiecutter/config.yaml
    ```
    and your project is rendered from the template with no prompts.

**Notes**
1. Values in the `default_context` block of the user config file are injected into **all** cookiecutter projects. So, after generating your first project template (but before you generate your second project template), you would need to replace these config file values with a new set of defaults that are specific to the second template to be rendered.

Hope this helps with your question.

**Additional Links**
1. [Relevant stackoverflow question](https://stackoverflow.com/a/60905956/4057186)

**Comment by jamesbraza** [comment]
@edesz yes the `--no-input` option closed the loop for programmatic access. :)

It seems `cookiecutter` supported it the whole time, case closed.  Also, thank you for the detailed comment!
